Source/WebCore/Modules/mediastream/ImageCapture.cpp

```diff
@@ -81,35 +81,42 @@ void ImageCapture::takePhoto(PhotoSettings&& settings, DOMPromiseDeferred<IDLInt
     });
 }
 
-void ImageCapture::getPhotoCapabilities(PhotoCapabilitiesPromise&& promise)
+void ImageCapture::getPhotoCapabilities(DOMPromiseDeferred<IDLDictionary<PhotoCapabilities>>&& promise)
 {
     auto identifier = LOGIDENTIFIER;
     ALWAYS_LOG(identifier);
 
-    if (m_track->readyState() == MediaStreamTrack::State::Ended) {
-        ERROR_LOG(identifier, "rejecting promise, track has ended");
-        promise.reject(Exception { ExceptionCode::InvalidStateError, "Track has ended"_s });
-        return;
-    }
+    m_track->getPhotoCapabilities()->whenSettled(RunLoop::main(), [this, protectedThis = Ref { *this }, promise = WTFMove(promise), identifier = WTFMove(identifier)] (auto&& result) mutable {
+        queueTaskKeepingObjectAlive(*this, TaskSource::ImageCapture, [this, promise = WTFMove(promise), result = WTFMove(result), protectedThis, identifier = WTFMove(identifier)] () mutable {
+            if (!result) {
+                ERROR_LOG(identifier, "rejecting promise: ", result.error().message());
+                promise.reject(WTFMove(result.error()));
+                return;
+            }
 
-    m_track->getPhotoCapabilities(WTFMove(promise));
+            ALWAYS_LOG(identifier, "resolving promise");
+            promise.resolve(WTFMove(result.value()));
+        });
+    });
 }
 
-void ImageCapture::getPhotoSettings(PhotoSettingsPromise&& promise)
+void ImageCapture::getPhotoSettings(DOMPromiseDeferred<IDLDictionary<PhotoSettings>>&& promise)
 {
     auto identifier = LOGIDENTIFIER;
     ALWAYS_LOG(identifier);
 
-    if (m_track->readyState() == MediaStreamTrack::State::Ended) {
-        // https://w3c.github.io/mediacapture-image/#ref-for-dom-imagecapture-getphotosettings②
-        // If the readyState of track provided in the constructor is not live, return a promise
-        // rejected with a new DOMException whose name is InvalidStateError, and abort these steps.
-        ERROR_LOG(identifier, "rejecting promise, track has ended");
-        promise.reject(Exception { ExceptionCode::InvalidStateError, "Track has ended"_s });
-        return;
-    }
+    m_track->getPhotoSettings()->whenSettled(RunLoop::main(), [this, protectedThis = Ref { *this }, promise = WTFMove(promise), identifier = WTFMove(identifier)] (auto&& result) mutable {
+        queueTaskKeepingObjectAlive(*this, TaskSource::ImageCapture, [this, promise = WTFMove(promise), result = WTFMove(result), protectedThis, identifier = WTFMove(identifier)] () mutable {
+            if (!result) {
```

Comment:
you should be able to do directly `promise.settle(WTFMove(result));`

Reply:
I could, but we need to check the result anyway because I want to log success or failure